### DISCUSSION TITLE
Ability to pass functions in state object with useNavigate()

### DISCUSSION THREAD

**Discussion by kghosein** [remix-run/react-router#10335] [Proposals]
Recently I encountered a situation where I had to pass a function in state object, but it turns out that you can only pass serializable data in state object, so I had to pass the function as a string and dangerously invoke it with eval() in the receiving component. 

Why functions cannot be passed in the state object and what are the implications involved with passing functions? Does it has been ignored deliberately?

**Comment by brophdawg11** [comment]
That's due to the underlying `history` API: https://developer.mozilla.org/en-US/docs/Web/API/History_API/Working_with_the_History_API#the_pushstate_method

> The state object can be anything that can be serialized

<img width="724" alt="Screenshot 2023-04-13 at 9 29 11 AM" src="https://user-images.githubusercontent.com/1609022/231773893-11506918-ee8e-44a0-bc03-88a0dfe8c93a.png">

**Reply by kghosein** [reply]
Alright, but any suggestions if eval() or Function() constructor can avoided or any work around (because passing function is inevitable)? Every one on the internet talks about the security risks involved with eval() or Function() constructor but never talks about any solutions especially in reference to React. 

MDN do suggests [using indirect eval()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval#never_use_eval!) and there's also a stackoverflow thread (but without any solution that can actually be picked).

**Reply by brophdawg11** [reply]
It depends on the use case but personally I'd try to find a way to avoid needing a function and instead pass data through `state`
